--- a/channels/chan_sip.c
+++ b/channels/chan_sip.c
@@ -6585,9 +6585,9 @@ static int sip_indicate(struct ast_chann
 				ast_set_flag(&p->flags[2], SIP_PAGE3_REMOTE_HOLD_STATUS);
 				ast_debug(1, "--- HOLD - Activated remote hold \n");
 				if (T38_ENABLED == p->t38.state) {
-					transmit_reinvite_with_sdp(p, TRUE, TRUE);
+					transmit_reinvite_with_sdp(p, TRUE, FALSE);
 				} else {
-					transmit_reinvite_with_sdp(p, FALSE, TRUE);
+					transmit_reinvite_with_sdp(p, FALSE, FALSE);
 				}
 			}
 		} else {
@@ -6601,9 +6601,9 @@ static int sip_indicate(struct ast_chann
 			ast_clear_flag(&p->flags[2], SIP_PAGE3_REMOTE_HOLD_STATUS);
 			ast_debug(1, "--- HOLD - cleared remote hold \n");
 			if (T38_ENABLED == p->t38.state) {
-				transmit_reinvite_with_sdp(p, TRUE, TRUE);
+				transmit_reinvite_with_sdp(p, TRUE, FALSE);
 			} else {
-				transmit_reinvite_with_sdp(p, FALSE, TRUE);
+				transmit_reinvite_with_sdp(p, FALSE, FALSE);
 			}
 		} else {
 			ast_rtp_instance_update_source(p->rtp);
