--- a/flash_eraseall.c	2009-03-14 02:09:35.000000000 +0100
+++ b/flash_eraseall.c	2012-09-13 14:06:45.000000000 +0200
@@ -47,6 +47,7 @@
 static const char *mtd_device;
 static int quiet;		/* true -- don't output progress */
 static int jffs2;		// format for jffs2 usage
+static int broken_cfe;	// ignore oob layout and just write the cleanmarker at the start
 
 static void process_options (int argc, char *argv[]);
 void show_progress (mtd_info_t *meminfo, erase_info_t *erase);
@@ -162,8 +163,13 @@
 		if (isNAND) {
 			struct mtd_oob_buf oob;
 			oob.ptr = (unsigned char *) &cleanmarker;
-			oob.start = erase.start + clmpos;
-			oob.length = clmlen;
+			if (broken_cfe) {
+				oob.start = erase.start;
+				oob.length = 8;
+			} else {
+				oob.start = erase.start + clmpos;
+				oob.length = clmlen;
+			}
 			if (ioctl (fd, MEMWRITEOOB, &oob) != 0) {
 				fprintf(stderr, "\n%s: %s: MTD writeoob failure: %s\n", exe_name, mtd_device, strerror(errno));
 				continue;
@@ -198,11 +204,12 @@
 
 	for (;;) {
 		int option_index = 0;
-		static const char *short_options = "jq";
+		static const char *short_options = "jqb";
 		static const struct option long_options[] = {
 			{"help", no_argument, 0, 0},
 			{"version", no_argument, 0, 0},
 			{"jffs2", no_argument, 0, 'j'},
+			{"broken", no_argument, 0, 'b'},			
 			{"quiet", no_argument, 0, 'q'},
 			{"silent", no_argument, 0, 'q'},
 
@@ -229,6 +236,9 @@
 			case 'q':
 				quiet = 1;
 				break;
+			case 'b':
+				broken_cfe = 1;
+				break;
 			case 'j':
 				jffs2 = 1;
 				break;
@@ -264,6 +274,7 @@
 			"Erases all of the specified MTD device.\n"
 			"\n"
 			"  -j, --jffs2    format the device for jffs2\n"
+			"  -b, --broken   don't respect oob layout, needed for some Broadcom cfe\n"
 			"  -q, --quiet    don't display progress messages\n"
 			"      --silent   same as --quiet\n"
 			"      --help     display this help and exit\n"
