--- a/networking/udhcp/common.c
+++ b/networking/udhcp/common.c
@@ -64,6 +64,7 @@ const struct dhcp_optflag dhcp_optflags[] = {
 	{ OPTION_STATIC_ROUTES                    , 0xf9 }, /* DHCP_MS_STATIC_ROUTES */
 	{ OPTION_6RD | OPTION_REQ                 , 0xd4 }, /* DHCP_6RD (RFC)     */
 	{ OPTION_6RD                              , 0x96 }, /* DHCP_6RD (Comcast) */
+	{ OPTION_STRING				  , 0xf1 }, /* Option 241 private option */
 	{ OPTION_STRING                           , 0xfc }, /* DHCP_WPAD          */
 
 	/* Options below have no match in dhcp_option_strings[],
@@ -135,6 +136,7 @@ const char dhcp_option_strings[] ALIGN1 =
 	"msstaticroutes""\0"/* DHCP_MS_STATIC_ROUTES */
 	"ip6rd" "\0"       /* DHCP_6RD (RFC)      */
 	"ip6rd" "\0"       /* DHCP_6RD (Comcast)  */
+	"privopt241" "\0"  /* Option 241 private option */
 	"wpad" "\0"        /* DHCP_WPAD           */
 	;
 
