--- a/networking/udhcp/common.c	2015-05-12 16:38:28.947608922 +0200
+++ b/networking/udhcp/common.c	2015-05-11 15:24:40.525624792 +0200
@@ -58,8 +58,8 @@ const struct dhcp_optflag dhcp_optflags[
 	{ OPTION_STATIC_ROUTES    | OPTION_REQ     , 0x79 }, /* DHCP_STATIC_ROUTES */
 	{ OPTION_STRING   | OPTION_REQ             , 0x80 }, /* TFPT Server IP address. Will be a HTTP URL IN IUP       */
 #if ENABLE_FEATURE_UDHCP_8021Q
-	{ OPTION_U16  | OPTION_REQ                 , 0x84 }, /* DHCP_VLAN_ID       */
-	{ OPTION_U8   | OPTION_REQ                 , 0x85 }, /* DHCP_VLAN_PRIORITY */
+	{ OPTION_STRING                            , 0x84 }, /* DHCP_VLAN_ID       */
+	{ OPTION_STRING                            , 0x85 }, /* DHCP_VLAN_PRIORITY */
 #endif
 	{ OPTION_STATIC_ROUTES                    , 0xf9 }, /* DHCP_MS_STATIC_ROUTES */
 	{ OPTION_6RD | OPTION_REQ                 , 0xd4 }, /* DHCP_6RD (RFC)     */
