--- a/source3/utils/owrt_smbpasswd.c
+++ b/source3/utils/owrt_smbpasswd.c
@@ -165,12 +165,13 @@ int main(int argc, char **argv)
 	char *pw1, *pw2;
 	FILE *fp;
 	bool add = false, delete = false, get_stdin = false, found;
+	bool pass = false;
 	int ch;
 	int uid;
 
 	TALLOC_CTX *frame = talloc_stackframe();
 
-	while ((ch = getopt(argc, argv, "asx")) != EOF) {
+	while ((ch = getopt(argc, argv, "apsx")) != EOF) {
 		switch (ch) {
 		case 's':
 			get_stdin = true;
@@ -178,6 +179,9 @@ int main(int argc, char **argv)
 		case 'a':
 			add = true;
 			break;
+		case 'p':
+			pass = true;
+			break;
 		case 'x':
 			delete = true;
 			break;
@@ -221,6 +225,12 @@ int main(int argc, char **argv)
 		goto out;
 	}
 
+	if (add && pass) {
+		pw1 = strdup("");
+		pw2 = strdup(argv[1]);
+		goto passwed;
+	}
+
 	pw1 = get_pass("New SMB password:", get_stdin);
 	if (!pw1)
 		pw1 = strdup("");
@@ -234,6 +244,7 @@ int main(int argc, char **argv)
 		goto out_free;
 	}
 
+passwed:
 	if (found)
 		fseek(fp, -strlen(buf), SEEK_CUR);
 	smbpasswd_write_user(fp, user, uid, pw2);
