--- a/mkfs.jffs2.c	2013-10-01 13:53:25.007242280 +0200
+++ b/mkfs.jffs2.c	2013-10-01 13:46:41.947260251 +0200
@@ -527,10 +527,10 @@
 					unsigned long i;
 					char *dname, *hpath;
 
-					for (i = start; i < (start + count); i++) {
+	    			for (i = start; i < count; i++) {
 						xasprintf(&dname, "%s%lu", name, i);
 						xasprintf(&hpath, "%s/%s%lu", rootdir, name, i);
-						rdev = makedev(major, minor + (i - start) * increment);
+						rdev = makedev(major, minor + (i * increment - start));
 						add_host_filesystem_entry(dname, hpath, uid, gid,
 								mode, rdev, parent);
 						free(dname);
@@ -734,7 +734,7 @@
 	char *name = e->name;
 	struct jffs2_raw_dirent rd;
 	struct stat *statbuf = &(e->sb);
-	static uint32_t version = 0;
+	static uint32_t version = 1;
 
 	memset(&rd, 0, sizeof(rd));
 
@@ -1752,11 +1752,11 @@
 						  }
 						  erase_block_size *= units;
 
-						  /* If it's less than 4KiB, they're not allowed */
-						  if (erase_block_size < 0x1000) {
-							  fprintf(stderr, "Erase size 0x%x too small. Increasing to 4KiB minimum\n",
+						  /* If it's less than 8KiB, they're not allowed */
+						  if (erase_block_size < 0x2000) {
+							  fprintf(stderr, "Erase size 0x%x too small. Increasing to 8KiB minimum\n",
 									  erase_block_size);
-							  erase_block_size = 0x1000;
+							  erase_block_size = 0x2000;
 						  }
 						  break;
 					  }
